### PR TITLE
vscode: migrate to request type api

### PR DESCRIPTION
**Pull request by Veetaha**
More type-safety to the god of type-safety.

### CODE REVIEW

File: editors/code/src/inlay_hints.ts [hunk 1]
```diff
@@ -39,16 +39,6 @@ export function activateInlayHints(ctx: Ctx) {
     void hintsUpdater.setEnabled(ctx.config.displayInlayHints);
 }
 
-interface InlayHintsParams {
-    textDocument: lc.TextDocumentIdentifier;
-}
-
-interface InlayHint {
-    range: vscode.Range;
```
Comment (Veetaha) [inline editors/code/src/inlay_hints.ts]: @SomeoneToIgnore what was the reason to declare `vscode.Range` instead of `lc.Range` here? It worked all the time since their shape is structurally the same, but we may not rely on this, so I explicitly added `protocol2CodeConverter.asRange()` call here...

File: editors/code/src/rust-analyzer-api.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/**
+ * This file mirrors `crates/rust-analyzer/src/req.rs` declarations.
+ */
+
+import { RequestType, TextDocumentIdentifier, Position, Range, TextDocumentPositionParams, Location, NotificationType, WorkspaceEdit } from "vscode-languageclient";
```
Comment (matklad) [inline editors/code/src/rust-analyzer-api.ts]: import * as lc?

File: editors/code/src/rust-analyzer-api.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/**
+ * This file mirrors `crates/rust-analyzer/src/req.rs` declarations.
```
Comment (matklad) [inline editors/code/src/rust-analyzer-api.ts]: it's unfortunate that the files are named differently, but I also actively dislike both names :)

File: editors/code/src/inlay_hints.ts [hunk 4]
```diff
@@ -151,20 +141,15 @@ class HintsUpdater {
...
+        const request = { textDocument: { uri: documentUri } };
 
-        return sendRequestWithRetry<InlayHint[]>(
-            this.ctx.client,
-            'rust-analyzer/inlayHints',
-            request,
-            tokenSource.token
-        )
+        return sendRequestWithRetry(this.ctx.client, ra.inlayHints, request, tokenSource.token)
             .catch(_ => null)
```
Comment (matklad) [inline editors/code/src/inlay_hints.ts]: Might as well swtich to `client.logFailedRequest` here and elsewhere?

File: editors/code/src/rust-analyzer-api.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ */
+
+import { RequestType, TextDocumentIdentifier, Position, Range, TextDocumentPositionParams, Location, NotificationType, WorkspaceEdit } from "vscode-languageclient";
+
+type Option<T> = null | T;
+type Vec<T> = T[];
+type FxHashMap<K extends PropertyKey, V> = Record<K, V>;
```
Comment (Veetaha) [inline editors/code/src/rust-analyzer-api.ts]: Cheatsheet for Rust -> TS transformations...

File: editors/code/src/rust-analyzer-api.ts (hunk 2 above)
Comment (Veetaha) [reply editors/code/src/rust-analyzer-api.ts]: Followed the same import pattern we have in `req.rs`

File: editors/code/src/rust-analyzer-api.ts (hunk 3 above)
Comment (kjeremy) [reply editors/code/src/rust-analyzer-api.ts]: As an aside I really dislike `req.rs`... it should probably be `custom_requests.rs` or something with only the imports required to support that file. We end up importing a lot from that file in handlers and conv.

Comment (Veetaha) [reply editors/code/src/rust-analyzer-api.ts]: The name is awesome, unless come up with better!

File: editors/code/src/inlay_hints.ts (hunk 4 above)
Comment (Veetaha) [reply editors/code/src/inlay_hints.ts]: There is no need to log failed request here, we just pass the output channel to `client` [constructor](https://github.com/microsoft/vscode-languageserver-node/blob/f425af9de46a0187adb78ec8a46b9b2ce80c5412/client/src/client.ts#L2585).

Repeating `client.logFailedRequest` everywhere would be very annoying. What if we add the function that wraps `sendRequest` and logs on failure or wrap the client with our own `RaClient`?
```ts
class RustAnalyzerClient {
    inner: lc.LanguageClient;

    async sendRequest(...params) {
        return this.inner.sendRequest(...params).catch(err => {
            this.inner.logFailedRequest(...);
            throw err;
        });
    }

    async sendRequestWithRetry(...params) { ... }
  
    get p2c() { return this.inner.protocol2CodeConverter; }
    get c2p() { return this.inner.code2ProtocolConverter; }
    /* stuff */ 
}
```

File: editors/code/src/rust-analyzer-api.ts (hunk 2 above)
Comment (Veetaha) [reply editors/code/src/rust-analyzer-api.ts]: Okay, done

File: editors/code/src/commands/on_enter.ts [hunk 6]
```diff
@@ -10,22 +10,15 @@ async function handleKeypress(ctx: Ctx) {
...
         ),
-    };
-    const change = await client.sendRequest<undefined | SourceChange>(
-        'rust-analyzer/onEnter',
-        request,
-    ).catch(
-        (_error: any) => {
-            // FIXME: switch to the more modern (?) typed request infrastructure
-            // client.logFailedRequest(OnEnterRequest.type, error);
-            return Promise.resolve(null);
```
Comment (Veetaha) [inline editors/code/src/commands/on_enter.ts]: FYI: @matklad  there is no need to wrap the return value into a promise in the `then(onfulfilled, onrejected)/catch(onrejected)` handlers.

File: editors/code/src/highlighting.ts [hunk 7]
```diff
@@ -45,13 +41,10 @@ export function activateHighlighting(ctx: Ctx) {
             const client = ctx.client;
             if (!client) return;
```
Comment (Veetaha) [inline editors/code/src/highlighting.ts]: We should get rid of these if statements, since the client is not nullable today...
But this deserves a separate PR.

File: editors/code/src/inlay_hints.ts (hunk 1 above)
Comment (SomeoneToIgnore) [reply editors/code/src/inlay_hints.ts]: No particular reason, pick the one that suits better and works still.

File: editors/code/src/highlighting.ts (hunk 7 above)
Comment (matklad) [reply editors/code/src/highlighting.ts]: yup, there's FIXME about this in main.ts
